File: ios/LocationEngine.swift
import Foundation
import CoreLocation

/**
 * Protocol for receiving location updates and errors from the LocationEngine.
 */
protocol LocationUpdateDelegate: AnyObject {
    func onLocationReceived(location: CLLocation)
    func onLocationError(errorCode: String, message: String)
}

/**
 * Default implementation for optional delegate method so existing code keeps compiling.
 */
extension LocationUpdateDelegate {
    func onLocationError(errorCode: String, message: String) {
        // No-op by default
    }
}

/**
 * iOS Location Engine that wraps CLLocationManager.
 * Provides high-accuracy location updates for live tracking with background support.
 *
 * Requirements: 3.1, 3.2
 */
class LocationEngine: NSObject, CLLocationManagerDelegate {

    private var locationManager: CLLocationManager!

    weak var delegate: LocationUpdateDelegate?

    override init() {
        super.init()
        let initOnMain = { [self] in
            self.locationManager = CLLocationManager()
            self.locationManager.delegate = self
        }
        if Thread.isMainThread {
            initOnMain()
        } else {
            DispatchQueue.main.sync(execute: initOnMain)
        }
    }

    // MARK: - Public Methods

    func requestAlwaysAuthorization() {
        if Thread.isMainThread {
            self.locationManager.requestAlwaysAuthorization()
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.locationManager.requestAlwaysAuthorization()
            }
        }
    }

    func getAuthorizationStatus() -> CLAuthorizationStatus {
        if #available(iOS 14.0, *) {
            return locationManager.authorizationStatus
        } else {
            return CLLocationManager.authorizationStatus()
        }
    }

    /**
     * Start receiving location updates with the specified interval and distance filter.
     * Uses kCLLocationAccuracyBest for best possible location accuracy.
     *
     * - Parameter intervalMs: The desired interval for location updates in milliseconds (used for time-based filtering at a higher level)
     * - Parameter distanceFilter: The minimum distance between updates in meters
     */
    func startLocationUpdates(intervalMs: Int, distanceFilter: Double) {
        startLocationUpdates(intervalMs: intervalMs, distanceFilter: distanceFilter, accuracy: kCLLocationAccuracyBest)
    }

    /**
     * Start receiving location updates with the specified interval, distance filter, and accuracy.
     * Allows configurable accuracy for scenarios like sleep mode where lower accuracy saves battery.
     *
     * - Parameter intervalMs: The desired interval for location updates in milliseconds (used for time-based filtering at a higher level)
     * - Parameter distanceFilter: The minimum distance between updates in meters
     * - Parameter accuracy: The desired location accuracy (e.g., kCLLocationAccuracyBest or kCLLocationAccuracyKilometer)
     */
    func startLocationUpdates(intervalMs: Int, distanceFilter: Double, accuracy: CLLocationAccuracy) {
        let setup = { [weak self] in
            guard let self = self else { return }
            self.locationManager.desiredAccuracy = accuracy
            self.locationManager.distanceFilter = distanceFilter
            self.locationManager.allowsBackgroundLocationUpdates = true
            self.locationManager.pausesLocationUpdatesAutomatically = false
            // Use otherNavigation so iOS does not throttle/suspend updates when device is stationary
            self.locationManager.activityType = .otherNavigation
            self.locationManager.showsBackgroundLocationIndicator = true
            self.locationManager.startUpdatingLocation()
        }
        if Thread.isMainThread {
            setup()
        } else {
            DispatchQueue.main.async(execute: setup)
        }
    }

    /**
     * Stop receiving location updates.
     */
    func stopLocationUpdates() {
        if Thread.isMainThread {
            locationManager.stopUpdatingLocation()
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.locationManager.stopUpdatingLocation()
            }
        }
    }

    // MARK: - CLLocationManagerDelegate

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        print("[LocationEngine] 📡 didUpdateLocations called — delegate alive=\(delegate != nil) lat=\(location.coordinate.latitude) ts=\(Int64(location.timestamp.timeIntervalSince1970 * 1000))")
        delegate?.onLocationReceived(location: location)
    }

    func locationManagerDidPauseLocationUpdates(_ manager: CLLocationManager) {
        print("[LocationEngine] ⚠️ CLLocationManager PAUSED location updates!")
    }

    func locationManagerDidResumeLocationUpdates(_ manager: CLLocationManager) {
        print("[LocationEngine] ✅ CLLocationManager RESUMED location updates")
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        let errorCode: String
        let message: String

        if let clError = error as? CLError {
            switch clError.code {
            case .denied:
                errorCode = "PERMISSION_DENIED"
                message = "Location updates failed: permission denied."
            case .locationUnknown:
                errorCode = "LOCATION_UNKNOWN"
                message = "Location temporarily unavailable: \(error.localizedDescription)"
            case .network:
                errorCode = "NETWORK_ERROR"
                message = "Location network error: \(error.localizedDescription)"
            default:
                errorCode = "LOCATION_ERROR"
                message = "Location update failed: \(error.localizedDescription)"
            }
        } else {
            errorCode = "LOCATION_ERROR"
            message = "Location update failed: \(error.localizedDescription)"
        }

        delegate?.onLocationError(errorCode: errorCode, message: message)
        print("[LocationEngine] \(errorCode): \(message)")
    }
}

File: android/src/main/java/com/livetracking/location/LocationEngine.kt
package com.livetracking.location

import android.content.Context
import android.location.Location
import android.os.Looper
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority

/**
 * Android Location Engine that wraps FusedLocationProviderClient.
 * Provides high-accuracy location updates for live tracking.
 *
 * Requirements: 2.2
 */
class LocationEngine(context: Context) {

    /**
     * Listener interface for receiving location updates from the engine.
     */
    interface LocationUpdateListener {
        fun onLocationReceived(location: Location)
    }

    private val fusedLocationClient: FusedLocationProviderClient =
        LocationServices.getFusedLocationProviderClient(context)

    private var locationListener: LocationUpdateListener? = null

    private val locationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
            val location = locationResult.lastLocation ?: return
            locationListener?.onLocationReceived(location)
        }
    }

    /**
     * Set the listener that will receive location updates.
     *
     * @param listener The listener to receive location callbacks
     */
    fun setLocationListener(listener: LocationUpdateListener) {
        this.locationListener = listener
    }

    /**
     * Start receiving location updates with the specified interval and distance filter.
     * Uses PRIORITY_HIGH_ACCURACY for best possible location accuracy.
     *
     * @param intervalMs The desired interval for location updates in milliseconds
     * @param distanceFilter The minimum distance between updates in meters
     * @throws SecurityException if location permissions are not granted
     */
    @Throws(SecurityException::class)
    fun startLocationUpdates(intervalMs: Long, distanceFilter: Float) {
        startLocationUpdates(intervalMs, distanceFilter, Priority.PRIORITY_HIGH_ACCURACY)
    }

    /**
     * Start receiving location updates with the specified interval, distance filter, and priority.
     *
     * @param intervalMs The desired interval for location updates in milliseconds
     * @param distanceFilter The minimum distance between updates in meters
     * @param priority The location request priority (e.g., Priority.PRIORITY_HIGH_ACCURACY or Priority.PRIORITY_LOW_POWER)
     * @throws SecurityException if location permissions are not granted
     */
    @Throws(SecurityException::class)
    fun startLocationUpdates(intervalMs: Long, distanceFilter: Float, priority: Int) {
        val locationRequest = LocationRequest.Builder(priority, intervalMs)
            .setMinUpdateDistanceMeters(distanceFilter)
            .setWaitForAccurateLocation(false)
            .build()

        fusedLocationClient.requestLocationUpdates(
            locationRequest,
            locationCallback,
            Looper.getMainLooper()
        )
    }

    /**
     * Stop receiving location updates and remove the location callback.
     */
    fun stopLocationUpdates() {
        fusedLocationClient.removeLocationUpdates(locationCallback)
    }
}
